server/types/uuid.go

```diff
@@ -0,0 +1,193 @@
+type UuidType struct{}
+
+var _ types.ExtendedType = UuidType{}
+
+// CollationCoercibility implements the types.ExtendedType interface.
+func (b UuidType) CollationCoercibility(ctx *sql.Context) (collation sql.CollationID, coercibility byte) {
+	return sql.Collation_binary, 5
+}
+
+// Compare implements the types.ExtendedType interface.
+func (b UuidType) Compare(v1 any, v2 any) (int, error) {
+	if v1 == nil && v2 == nil {
+		return 0, nil
+	} else if v1 != nil && v2 == nil {
+		return 1, nil
+	} else if v1 == nil && v2 != nil {
+		return -1, nil
+	}
+
+	ac, _, err := b.Convert(v1)
+	if err != nil {
+		return 0, err
+	}
+	bc, _, err := b.Convert(v2)
+	if err != nil {
+		return 0, err
+	}
+
+	ab := ac.(uuid.UUID)
+	bb := bc.(uuid.UUID)
+	return bytes.Compare(ab.GetBytesMut(), bb.GetBytesMut()), nil
+}
+
+// Convert implements the types.ExtendedType interface.
+func (b UuidType) Convert(val any) (any, sql.ConvertInRange, error) {
+	if val == nil {
+		return nil, sql.InRange, nil
+	}
+
+	switch val := val.(type) {
+	case string:
+		uuidVal, err := uuid.FromString(val)
+		if err != nil {
+			return nil, sql.OutOfRange, err
+		}
+		return uuidVal, sql.InRange, nil
+	case uuid.UUID:
+		return val, sql.InRange, nil
+	default:
+		return nil, sql.OutOfRange, sql.ErrInvalidType.New(b)
+	}
+}
+
+// Equals implements the types.ExtendedType interface.
+func (b UuidType) Equals(otherType sql.Type) bool {
+	if otherExtendedType, ok := otherType.(types.ExtendedType); ok {
+		return bytes.Equal(MustSerializeType(b), MustSerializeType(otherExtendedType))
+	}
+	return false
+}
+
+// FormatSerializedValue implements the types.ExtendedType interface.
+func (b UuidType) FormatSerializedValue(val []byte) (string, error) {
+	deserialized, err := b.DeserializeValue(val)
+	if err != nil {
+		return "", err
+	}
+	return b.FormatValue(deserialized)
+}
+
+// FormatValue implements the types.ExtendedType interface.
+func (b UuidType) FormatValue(val any) (string, error) {
+	if val == nil {
+		return "", nil
+	}
+	converted, _, err := b.Convert(val)
+	if err != nil {
+		return "", err
+	}
+	return converted.(uuid.UUID).String(), nil
+}
+
+// MaxSerializedWidth implements the types.ExtendedType interface.
+func (b UuidType) MaxSerializedWidth() types.ExtendedTypeSerializedWidth {
+	return types.ExtendedTypeSerializedWidth_64K
+}
+
+// MaxTextResponseByteLength implements the types.ExtendedType interface.
+func (b UuidType) MaxTextResponseByteLength(ctx *sql.Context) uint32 {
+	return 16
+}
+
+// Promote implements the types.ExtendedType interface.
+func (b UuidType) Promote() sql.Type {
+	return b
+}
+
+// SerializedCompare implements the types.ExtendedType interface.
+func (b UuidType) SerializedCompare(v1 []byte, v2 []byte) (int, error) {
+	if len(v1) == 0 && len(v2) == 0 {
+		return 0, nil
+	} else if len(v1) > 0 && len(v2) == 0 {
+		return 1, nil
+	} else if len(v1) == 0 && len(v2) > 0 {
+		return -1, nil
+	}
+
+	return bytes.Compare(v1, v2), nil
```

Comment:
Sanity check: does this work with the string ascii values correctly?

Reply:
It's not meant to. This is in `SerializedCompare`, which is different from `Compare` in that it only works on the serialized forms. If a string has made it in here, then something has gone horribly wrong somewhere else in the code.